metric.go

```diff
@@ -0,0 +1,476 @@
+package gosnowth
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+)
+
+// scanToken values represent individual tokens found by query scanners.
+type scanToken uint
+
+// scanToken values used by the scanners to identify tokens.
+const (
+	tokenEOF scanToken = iota
+	tokenIllegal
+	tokenB64 // Base64 encoding marker
+	tokenWS  // Whitespace
+	tokenOP  // Open parenthesis
+	tokenCP  // Close parenthesis
+	tokenOB  // Open bracket
+	tokenCB  // Close bracket
+	tokenColon
+	tokenComma
+	tokenQuote
+	tokenKeyword
+	tokenTagCat
+	tokenTagVal
+	tokenMetric
+	tokenStreamTag
+	tokenMeasurementTag
+)
+
+// tagType value represent whether a tag is a stream or measurment tag.
+type tagType uint
+
+const (
+	tagStreamTag = iota
+	tagMeasurementTag
+)
+
+// Tag values represent stream or measurment tags.
+type Tag struct {
+	Category string
+	Value    string
+}
+
+// MetricName values represent metric names.
+type MetricName struct {
+	CanonicalName   string
+	Name            string
+	StreamTags      []Tag
+	MeasurementTags []Tag
+}
+
+// NewMetricName creates and initializes a new metric name value.
+func NewMetricName() *MetricName {
+	return &MetricName{}
+}
+
+// metricScanner represents a lexical scanner for metric names.
+type metricScanner struct {
+	r *bufio.Reader
+}
+
+// newMetricScanner returns a new metric name scanner value.
+func newMetricScanner(r io.Reader) *metricScanner {
+	return &metricScanner{r: bufio.NewReader(r)}
+}
+
+// read reads the next rune from the bufferred reader.
+// Returns the rune(0) if an error occurs (or io.EOF is returned).
+func (ms *metricScanner) read() rune {
+	ch, _, err := ms.r.ReadRune()
+	if err != nil {
+		return rune(0) // EOF
+	}
+
+	return ch
+}
+
+// unread places the previously read rune back on the reader.
+func (ms *metricScanner) unread() error { return ms.r.UnreadRune() }
+
+// Scan returns the next token and literal value.
+func (ms *metricScanner) scan() (tok scanToken, lit string) {
+	ch := ms.read()
+	switch ch {
+	case rune(0):
+		return tokenEOF, ""
+	case '[':
+		return tokenOB, string(ch)
+	case ']':
+		return tokenCB, string(ch)
+	case '"':
+		return tokenQuote, string(ch)
+	case 'b':
+		return tokenB64, string(ch)
```

Comment:
I noticed in package tests, tokenQuote and tokenB64 are not getting exercised.  Is that expected?

Reply:
Good catch. The linter showed me that these are actually not being used in the parser, there is no case where scan() is used that these would be found. I have removed these cases.